### DISCUSSION TITLE
Method for using organic models?

### DISCUSSION THREAD

**Discussion by gvittrup** [google-deepmind/mujoco#242] [Asking for Help]
I am looking for a workflow that would allow the use of organic models in Unity for increased accuracy with motion capturing data and environmental game objects. A prime example I found for this was this [video](https://youtu.be/i0_OjDil0Fg) under the dm_control domain, though I am not sure if this was achieved by converting a file to MJCF, if it was an OBJ file, or if the mesh was rendered onto the frame displayed at 00:00:30 in the video linked above. 

Additionally, I noticed on the documentation its mention of skinning for visualizations, though I am not exactly confident this is what I am looking for nor does the documentation explain how to generate these files and if MuJoCo is dependent on a certain procedure.

Any point of reference or direction would be really appreciated!

**Comment by yuvaltassa** [comment]
This model lives [here](https://github.com/deepmind/dm_control/blob/main/dm_control/suite/dog.xml), assets in the `dog_assets` directory, and makes use of MuJoCo's [skin](https://mujoco.readthedocs.io/en/latest/XMLreference.html#asset-skin) asset type. Feel free to load it in `simulate` and play with it.

Import/export of skinned meshes to and from Unity is also possible, but currently not implemented.

**Comment by gvittrup** [comment]
What would the proper workflow be to generate SKN / MSH files? As per documentation, it says you can load custom binary files, though my test cases failed. I used the Godot engine, which exported my 3D model from the engine into a .msh, though despite other programs being able to read this file MuJoCo was unable to. 

Additionally, when talking about mapping vertices to bone subelements, does this require my model/armature to be rigged? I noticed in the dog_assets and the XML file that it referenced all of these bones but I am unsure if this is relative to the skinning of the mesh.

**Reply by yuvaltassa** [reply]
Hi, 

First note that MuJoCo's `.msh` format is a simple proprietary format defined [here](https://mujoco.readthedocs.io/en/latest/XMLreference.html#asset-mesh). I don't know what Godot is exporting, but it is surely not this.

Unfortunately making skins is not easy right now. It is certainly possible to do it using [PyMJCF](https://github.com/deepmind/dm_control/blob/main/dm_control/mjcf/README.md), which is how the dog skin was made. For this you need to compute vertex weights i.e., how much each vertex of your skin is affected by every body ("bone" in this context). 

If you want to see an example of a skin written directly in XML, try loading in `simulate` and saving as XML one of the composite objects that use skins, e.g. [this one](https://github.com/deepmind/mujoco/blob/main/model/composite/softbox.xml).

If/when you have successfully made your skin, the snippet below could help you turn it into a binary `.skn` file.

Let us know if you try any of this and get stuck.

```Python
# Convert skin into *.skn file accordong to
# http://mujoco.org/book/XMLreference.html#skin
f = open('dog_skin.skn', 'w+b')
nvert = skin.vertex.size//3
f.write(struct.pack('4i', nvert, nvert, skin.face.size//3, physics.model.nbody-1))
f.write(struct.pack(str(skin.vertex.size)+'f', *skin.vertex))
assert physics.model.mesh_texcoord.shape[0] == physics.bind(skin_msh).vertnum
f.write(struct.pack(str(2*nvert)+'f', *physics.model.mesh_texcoord.flatten()))
f.write(struct.pack(str(skin.face.size)+'i', *skin.face))
for bone in skin.bone:
  name_length = len(bone.body)
  assert name_length <= 40
  f.write(struct.pack(str(name_length)+'c', *[s.encode() for s in bone.body]))
  f.write((40-name_length)*b'\x00')
  f.write(struct.pack('3f', *bone.bindpos))
  f.write(struct.pack('4f', *bone.bindquat))
  f.write(struct.pack('i', bone.vertid.size))
  f.write(struct.pack(str(bone.vertid.size)+'i', *bone.vertid))
  f.write(struct.pack(str(bone.vertid.size)+'f', *bone.vertweight))
f.close()
```

**Reply by gvittrup** [reply]
Thank you for your continued support!

My approach will be writing a parser that converts the text document containing vertex and texture coordinate information of my model into a skin asset specified in an XML file and then using the above code to convert that into a binary skin file. I will be using the vertex coordinates to create the bodies, geom's, and joints.

Regarding the skin asset type, as it was created in the soft box composite object that you linked, what do the vertex and face components reference? I am confused as to why vertex only has a value of 0, and why face contains numbers up to the maximum vertices count of the composite object (293), but lists them in varying order and repetition. 

Additionally, any thoughts regarding my approach would be greatly appreciated.

Thanks again.

**Reply by yuvaltassa** [reply]
Perhaps you didn't find the [relevant section](https://mujoco.readthedocs.io/en/latest/XMLreference.html#asset-skin) explaining how skins work?

Please read that and let me know what questions remain?

**Reply by gvittrup** [reply]
My apologies - the slight discrepancy in naming derailed me.

The issue I have been struggling with since then is regarding vertex weights. Under the documentation, it lists that for each bone, it has vertex_index[vertex_count] and vertex_weight[vertex_count]. I recognized this in the composite soft_box model, as it has 294 vertices, and 294 bones, each with a vertex weight of 1. My model has 372 vertices, though I am receiving an error in the engine that my vertices must have positive total weight in skin. I can only break this error in the engine if I change the vertex weight in my skin to a ridiculously high number (i.e. 999999999999), to which it then creates an error saying that my vertex count is too high. I can replicate these errors on the soft_box model if I add or remove a single digit from its vertex weights, to which I understand the cause being the disparity in vertex count. What could be causing the engine to assume my vertex weight/vertex count is so ridiculously high?

**Reply by yuvaltassa** [reply]
Hi @gvittrup,

Could you try to make a minimal model that we could load and look at? it's difficult to diagnose your problem like this. Perhaps look at [this model](https://github.com/deepmind/mujoco/blob/main/test/engine/testdata/skingroup.xml) with an explicit XML skin, and it will clarify things for you?

**Reply by gvittrup** [reply]
Yuval,

Thanks for that example. I broke my model down to 32 vertices for testing and followed that syntax exactly, and I was able to get past that error and finally render my model as a skin.
